agent/taskresource/volume.go

```diff
@@ -0,0 +1,150 @@
+
+	return &VolumeResource{
+		Name: name,
+		Mountpoint: mountPoint,
+		Driver: driver,
+		DriverOptions: driverOptions,
+		Labels: labels,
+	}
+}
+
+// SetDesiredStatus safely sets the desired status of the resource
+func (vol *VolumeResource) SetDesiredStatus(status VolumeStatus) {
+	vol.lock.Lock()
+	defer vol.lock.Unlock()
+
+	vol.desiredStatusUnsafe = status
+}
+
+// GetDesiredStatus safely returns the desired status of the task
+func (vol *VolumeResource) GetDesiredStatus() VolumeStatus {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.desiredStatusUnsafe
+}
+
+// SetKnownStatus safely sets the currently known status of the resource
+func (vol *VolumeResource) SetKnownStatus(status VolumeStatus) {
+	vol.lock.Lock()
+	defer vol.lock.Unlock()
+
+	vol.knownStatusUnsafe = status
+}
+
+// GetKnownStatus safely returns the currently known status of the task
+func (vol *VolumeResource) GetKnownStatus() VolumeStatus {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.knownStatusUnsafe
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (vol *VolumeResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	vol.lock.Lock()
+	defer vol.lock.Unlock()
+
+	vol.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (vol *VolumeResource) GetCreatedAt() time.Time {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.createdAt
+}
+
+// volumeResourceJSON duplicates VolumeResource fields, only for marshalling and unmarshalling purposes
+type volumeResourceJSON struct {
+	Name          string            `json:"Name"`
+	Mountpoint    string            `json:"MountPoint"`
+	Driver        string            `json:"Driver"`
+	DriverOptions map[string]string `json:"DriverOptions"`
+	Labels        map[string]string `json:"Labels"`
+	DesiredStatus *VolumeStatus `json:"DesiredStatus"`
+	KnownStatus   *VolumeStatus `json:"KnownStatus"`
+}
+
+// MarshalJSON marshals VolumeResource object using duplicate struct VolumeResourceJSON
+func (vol *VolumeResource) MarshalJSON() ([]byte, error) {
+	if vol == nil {
+		return nil, nil
+	}
+	return json.Marshal(volumeResourceJSON{
+		vol.Name,
+		vol.Mountpoint,
+		vol.Driver,
+		vol.DriverOptions,
+		vol.Labels,
+		func() *VolumeStatus { desiredState := vol.GetDesiredStatus(); return &desiredState }(),
+		func() *VolumeStatus { knownState := vol.GetKnownStatus(); return &knownState }(),
+	})
+}
+
+// UnmarshalJSON unmarshals VolumeResource object using duplicate struct VolumeResourceJSON
+func (vol *VolumeResource) UnmarshalJSON(b []byte) error {
+	temp := &volumeResourceJSON{}
+
+	if err := json.Unmarshal(b, &temp); err != nil {
+		return err
+	}
+
+	vol.Name = temp.Name
+	vol.Mountpoint = temp.Mountpoint
+	vol.Driver = temp.Driver
+	vol.DriverOptions = temp.DriverOptions
+	vol.Labels = temp.Labels
+	if temp.DesiredStatus != nil {
+		vol.SetDesiredStatus(*temp.DesiredStatus)
```

Comment:
Similar to above, can you check if either of `aws.IntVal(temp.DesiredStatus)` or `aws.Int32Val(temp.DesiredStatus)` work here? You should be able to reuse one of those